File: f77/build08/term.f
C
        Subroutine Termination_Conditions(Quit)
C
C       This module determines if the run termination conditions have been met
C
C       Inputs:
C           X_bt_i  - X position of the BCS w.r.t. the target, expressed in the ICS [m]
C           Xd_bt_i - X velocity of the BCS w.r.t. the target, expressed in the ICS
C                             [m/sec]
C           Z_bt_i  - Z position of the BCS w.r.t. the target, expressed in the ICS [m]
C           Zd_bt_i - Z velocity of the BCS w.r.t. the target, expressed in the ICS
C                             [m/sec]
C
C       Outputs:
C           Quit - Stop simulation run [Boolean]
C
C       Internal variables and constants:
C           Time_To_Go - Time until closest approach [sec]
C
C       Declare global common and assign variable locations
C
      Include '../driver/global.inc'
      Include '../driver/sysvars.inc'
      Logical Quit
      Equivalence (Real_Array(602), X_bt_i)
      Equivalence (Real_Array(603), Z_bt_i)
      Equivalence (Real_Array(604), Xd_bt_i)
      Equivalence (Real_Array(605), Zd_bt_i)
C
C       Begin consition check:
C
C       Evaluate closest approach time and miss distance
C
      Time_To_Go = -(X_bt_i*Xd_bt_i + Z_bt_i*Zd_bt_i)/
     &              (Xd_bt_i**2 + Zd_bt_i**2)
      If (Time_To_Go .lt. 0.0e0) Then
         Call Miss
         Quit = .TRUE.
      End If
C
      Return
      End

      Subroutine Miss
C
C       Miss distance calculation routine
C
C       Inputs:
C           X_bt_i  - X position of the BCS w.r.t. the target, expressed in the ICS [m]
C           Xd_bt_i - X velocity of the BCS w.r.t. the target, expressed in the ICS
C                             [m/sec]
C           Z_bt_i  - Z position of the BCS w.r.t. the target, expressed in the ICS [m]
C           Zd_bt_i - Z velocity of the BCS w.r.t. the target, expressed in the ICS
C                             [m/sec]
C
C       Outputs:
C           DtMiss  - Time step to closest approach [sec]
C           RMiss   - Missile miss distance [m]
C
C       Internal variables and constants:
C           None.
C
C       Declare global common and assign variable locations

      Include '../driver/global.inc'
      Include '../driver/sysvars.inc'
      Equivalence (Real_Array(602), X_bt_i)
      Equivalence (Real_Array(603), Z_bt_i)
      Equivalence (Real_Array(604), Xd_bt_i)
      Equivalence (Real_Array(605), Zd_bt_i)
C
C       Begin miss distance calculation:
C
C       Evaluate closest approach time and miss distance
C
      DtMiss = (X_bt_i*Xd_bt_i + Z_bt_i*Zd_bt_i)/
     &         (Xd_bt_i**2     + Zd_bt_i**2)
      RMiss = Sqrt((X_bt_i - Xd_bt_i*DtMiss)**2 +
     &             (Z_bt_i - Zd_bt_i*DtMiss)**2)
      Open(Unit = 98, File = 'miss.dat')
      Write(98, *) 'RMiss = ',RMiss
      Write(98, *) 'DtMiss = ',Dtmiss
      Close(98)
C
      Return
      End

File: compcert/build08/term.c
#include "term.h"
#include "../driver/global.h"
#include "../driver/sysvars.h"
#include <stdio.h>
#include <math.h>

#define x_bt_i  real_array[602].r
#define z_bt_i  real_array[603].r
#define xd_bt_i real_array[604].r
#define zd_bt_i real_array[605].r

BOOL termination_conditions(BOOL quit) {
  // This module determines if the run termination conditions have been met
  //
  // Inputs:
  //     X_bt_i  - X position of the BCS w.r.t. the target, expressed in the ICS [m]
  //     Xd_bt_i - X velocity of the BCS w.r.t. the target, expressed in the ICS
  //                       [m/sec]
  //     Z_bt_i  - Z position of the BCS w.r.t. the target, expressed in the ICS [m]
  //     Zd_bt_i - Z velocity of the BCS w.r.t. the target, expressed in the ICS [m/sec]
  //
  // Outputs:
  //     Quit - Stop simulation run [Boolean]
  //
  // Internal variables and constants:
  //     Time_To_Go - Time until closest approach [sec]
  //
  // Declare global common and assign variable locations
  // Begin consition check:
  // Evaluate closest approach time and miss distance
  REAL time_to_go;
  time_to_go = -(x_bt_i*xd_bt_i + z_bt_i*zd_bt_i)/(pow(xd_bt_i,2) + pow(zd_bt_i,2));
  if (time_to_go < 0.0) {
    miss();
    quit = TRUE;
  }
  return quit;
}

void miss() {
  // Miss distance calculation routine
  //
  // Inputs:
  //     X_bt_i  - X position of the BCS w.r.t. the target, expressed in the ICS [m]
  //     Xd_bt_i - X velocity of the BCS w.r.t. the target, expressed in the ICS [m/sec]
  //     Z_bt_i  - Z position of the BCS w.r.t. the target, expressed in the ICS [m]
  //     Zd_bt_i - Z velocity of the BCS w.r.t. the target, expressed in the ICS [m/sec]
  //
  // Outputs:
  //     DtMiss  - Time step to closest approach [sec]
  //     RMiss   - Missile miss distance [m]
  REAL dtmiss,rmiss;
  FILE *missdat;
  // Begin miss distance calculation:
  // Evaluate closest approach time and miss distance
  dtmiss = (x_bt_i*xd_bt_i + z_bt_i*zd_bt_i)/(pow(xd_bt_i,2) + pow(zd_bt_i,2));
  rmiss  = sqrt(pow(x_bt_i - xd_bt_i*dtmiss,2) + pow(z_bt_i - zd_bt_i*dtmiss,2));
  missdat = fopen("miss.dat", "w");
  fprintf(missdat," RMiss = %f", rmiss);
  fprintf(missdat," DtMiss = %f", dtmiss);
  fclose(missdat);
}
